src/v/storage/disk_log_impl.cc

```diff
@@ -557,43 +557,71 @@ model::term_id disk_log_impl::term() const {
     }
     return _segs.back()->offsets().term;
 }
+static offset_stats empty_offsets(model::offset start_offset) {
```

Comment:
@mmaslankaprv given how we've updated the segment appender, what do you think of the strategy to flush before creating the new segment? - fdatasync should finish fast. what will tkae a while is truncating + closing the file, etc since it requires metadata synchronization. why not simply flush() before rolling?, and measuring that impact rather than updating the rest of the offset manipulation on the tree.

Reply:
Last, i think there is a 3rd strategy. so let me enumerate

1. the patch presented here
2. we flush before rolling - keep the rest of the code the same
3. we wait on the read for the flush to finish() -- 

 
**(3) is preferred** IMO. - basically we just need to add intelligence, that backpressures the read, but that way we don't block a group of writers. 

My fear w/ this manipulation is that I still am not sure the changes cover all use cases, this patch is very sensitive to the rest of the storage engine. 

Thoughts on (2) and (3) ?

Reply:
Those are nice ideas, I have some doubts about them. For both (2) & (3) we would still need to change `disk_log_impl::offsets()` method. Even if we would wait for flush (on closing segment) to finish before read,  we can not set `committed_offset` in the newly created segment. For `acks=1` we will defer flushing of that first batch appended to new segment. Additionally we already observe increased latency when segment is rolling it contributes to overall tail latency. I think that adding latency to segment roll may impact our tail latency results.

Reply:
Ok. I need to think more tom on this patch.

Reply:
@mmaslankaprv is this still the case w/ the latest patch from the performance improvements.  - also question, did we run this through punisher. 

I think what I am missing to understand this patch is a code-comment in the offset struct that explains the relationships of each offset with each other and when things get updated and when they don't get updated. 

for example, this doesn't cover the case where 

1. you roll an old segment with a large to-write backlog
2. you create a new segment & flush immediately
3. the new segment finishes the write/flush update *before* the old segment's data was flushed

I think i'm OK not advancing the commit offset, but the flush on the latest segment has to make sure the prev segment was flushed before 

I think we need a honey badger test here almost to prove things do what we expect them to do. i.e.: sleep the last write on the prev segment for 1 second and flush+read on the new segment 10 times. all the reads need to suspend until the 1 second of the prev segment finishes the flush.

Reply:
1. Segment roll always introduces latency, even with the most recent performance changes.
2. Yes, I am running it right now together with the most recent performance changes

I will add comments, I think we are missing the mechanism described in the last part of this sentence, i.e. we do not wait for the flush of prev segment when flushing the one that is newly created.

Reply:
@senior7515 together with @dotnwat we came up with idea of splitting flush fixes and offset fixes. Here is an issue: #1564

Reply:
fantastic. this change is very hard to read for me, and I wrote part of it. haha. soooo subtle.